src/oneid/jwts.py

```
@@ -0,0 +1,373 @@
+            logger.debug('invalid "typ" in header: %s', header)
+            return False
+
+        if 'alg' not in header or header['alg'] != 'ES256':
+            logger.debug('invalid "alg" in header: %s', header)
+            return False
+
+    logger.debug('returning %s', header)
+    return header
+
+
+def _verify_claims(payload):
+    try:
+        claims = json.loads(payload)
+        now = int(time.time())
+
+        if 'exp' in claims and (int(claims['exp']) + TOKEN_EXPIRATION_LEEWAY_SEC) < now:
+            logger.warning('Expired token, exp=%s, now=%s', claims['exp'], now)
+            return None
+
+        if 'nbf' in claims and (int(claims['nbf']) - TOKEN_NOT_BEFORE_LEEWAY_SEC) > now:
+            logger.warning('Early token, nbf=%s, now=%s', claims['nbf'], now)
+            return None
+
+        if 'jti' in claims and not utils.verify_and_burn_nonce(claims['jti']):
+            logger.warning('Invalid nonce: %s', claims['jti'])
+            return None
+
+        return claims
+
+    except:
+        logger.debug('unknown error verifying payload: %s', payload, exc_info=True)
+        return None
+
+
+def _verify_jws_signature(payload, keypair_map, signature):
+    # only use non-protected 'header' field for debugging
+    header = _verify_jose_header(
+        utils.to_string(utils.base64url_decode(signature['protected'])),
+        strict_jwt=False,
+    )
+
+    if not header:
+        return False
+
+    if 'kid' not in header:
+        logger.warning('invalid header in signature, missing "kid": %s', header)
+        return False
+
+    keypair = keypair_map.get(header['kid'])
+
+    if not keypair:
+        logger.warning('no keypair given for signature: %s', header)
+        return False
+
+    try:
+        keypair.verify('.'.join([signature['protected'], payload]), signature['signature'])
+    except:
+        logger.debug('invalid signature: %s', header)
+        return False
```

Instead of returning a `bool` it should raise an `InvalidSignature`.